core/consensus/src/adapter.rs

```diff
@@ -390,6 +436,199 @@ where
+                    )
+                    .into());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn verify_proof(
+        self: &Self,
+        ctx: Context,
+        block: Block,
+        proof: Proof,
+    ) -> ProtocolResult<()> {
+        // the block 0 has no proof, which is consensus-ed by community, not by chain
+
+        if block.header.height == 0 {
+            return Ok(());
+        };
+
+        if block.header.height != proof.height {
+            return Err(ConsensusError::VerifyProof(
+                block.header.height,
+                HeightMismatch(block.header.height, proof.height),
+            )
+            .into());
+        }
+
+        let blockhash = Hash::digest(block.clone().encode_fixed()?);
+
+        if blockhash != proof.block_hash {
+            return Err(ConsensusError::VerifyProof(block.header.height, HashMismatch).into());
+        }
+
+        let metadata = self.get_metadata(
+            ctx,
+            block.header.state_root.clone(),
+            block.header.height,
+            block.header.timestamp,
+        )?;
+
+        let mut authority_list = metadata
+            .verifier_list
+            .iter()
+            .map(|v| Node {
+                address:        v.address.as_bytes(),
+                propose_weight: v.propose_weight,
+                vote_weight:    v.vote_weight,
+            })
+            .collect::<Vec<Node>>();
+
+        let signed_voters = extract_voters(&mut authority_list, &proof.bitmap).map_err(|_| {
+            log::error!("[consensus] extract_voters fails, bitmap error");
+            ConsensusError::VerifyProof(block.header.height, BitMap)
+        })?;
+
+        let vote_hash = self.crypto.hash(protocol::Bytes::from(rlp::encode(&Vote {
+            height:     proof.height,
+            round:      proof.round,
+            vote_type:  VoteType::Precommit,
+            block_hash: proof.block_hash.as_bytes(),
+        })));
+
+        self.verify_proof_signature(
+            block.header.height,
+            vote_hash,
+            proof.signature,
+            signed_voters.clone(),
+        )?;
+
+        let weight_map = authority_list
+            .iter()
+            .map(|node| (node.address.clone(), node.vote_weight))
+            .collect::<HashMap<overlord::types::Address, u32>>();
+
+        self.verity_proof_weight(block.header.height, weight_map, signed_voters)?;
+
+        Ok(())
+    }
+
+    fn verify_proof_signature(
+        &self,
+        block_height: u64,
+        vote_hash: Bytes,
+        aggregated_signature_bytes: Bytes,
+        signed_voters: Vec<Bytes>,
+    ) -> ProtocolResult<()> {
+        // check sig
+        if self
+            .crypto
+            .verify_aggregated_signature(aggregated_signature_bytes, vote_hash, signed_voters)
+            .is_err()
+        {
+            return Err(ConsensusError::VerifyProof(block_height, Signature).into());
+        }
+        Ok(())
+    }
```

Comment:
The validators list might be changed, on account of the proof's height is equal to `current_overlord_height - 1`, it should get metadata before verify the aggregated signature.

Reply:
fixed